module-apps/apps-common/ApplicationCommon.cpp

```diff
@@ -530,56 +540,81 @@ namespace app
-                }
-                setActiveWindow(msg->getWindowName());
-            }
-            LOG_DEBUG("Current window: %s vs %s", getCurrentWindow()->getName().c_str(), msg->getWindowName().c_str());
-            getCurrentWindow()->handleSwitchData(switchData.get());
+        LOG_DEBUG("Current window: %s vs %s", getCurrentWindow()->getName().c_str(), msg->getWindowName().c_str());
+        // THIS IS IN TEST!
+        const auto &[name, data] = msg->getSwitchData();
+        pushWindow(name, data);
+        getCurrentWindow()->handleSwitchData(switchData.get());
 
-            auto ret = dynamic_cast<gui::SwitchSpecialChar *>(switchData.get());
-            if (ret != nullptr && ret->type == gui::SwitchSpecialChar::Type::Response) {
-                auto text = dynamic_cast<gui::Text *>(getCurrentWindow()->getFocusItem());
-                if (text != nullptr) {
-                    text->addText(ret->getDescription());
-                    refreshWindow(gui::RefreshModes::GUI_REFRESH_FAST);
-                    return sys::msgHandled();
-                }
-            }
-            getCurrentWindow()->onBeforeShow(msg->getCommand(), switchData.get());
+        if (handleUpdateTextRefresh(switchData.get())) {
+            return sys::msgHandled();
+        }
+        getCurrentWindow()->onBeforeShow(msg->getCommand(), switchData.get());
+
+        /// This is kind of gentle hack on our stateless Application Window management switching
+        /// if we are not requesting for popup and we cant handle next popup - just refresh the window
+        /// else -> popup will refresh it for us
+        if (not tryShowPopup()) {
+            LOG_INFO("REFRESHIG!");
             refreshWindow(gui::RefreshModes::GUI_REFRESH_DEEP);
         }
         else {
-            LOG_ERROR("No such window: %s", msg->getWindowName().c_str());
+            LOG_INFO("REFRESHIG we LOST REFRESH AS WE WILL REFRESH WITH POPUP");
         }
         return sys::msgHandled();
     }
 
+    bool ApplicationCommon::handleUpdateTextRefresh(gui::SwitchData *data)
+    {
+        auto ret = dynamic_cast<gui::SwitchSpecialChar *>(data);
+        if (ret != nullptr && ret->type == gui::SwitchSpecialChar::Type::Response) {
+            auto text = dynamic_cast<gui::Text *>(getCurrentWindow()->getFocusItem());
+            if (text != nullptr) {
+                text->addText(ret->getDescription());
+                refreshWindow(gui::RefreshModes::GUI_REFRESH_FAST);
+                return true;
+            }
+        }
+        return false;
+    }
+
     sys::MessagePointer ApplicationCommon::handleUpdateWindow(sys::Message *msgl)
     {
         auto msg = static_cast<AppUpdateWindowMessage *>(msgl);
-
-        if (windowsFactory.isRegistered(msg->getWindowName()) && isCurrentWindow(msg->getWindowName())) {
+        windowsStack().logStack(std::cout);
+        auto have_builder      = windowsFactory.isRegistered(msg->getWindowName());
```

Comment:
fix style. haveBuilder and isCurrentWindow

Reply:
also probably not haveBuilder but isRegistered?

Reply:
i get what you mean about `have_builder` but what about `isRegistered` ?